tidy3d/plugins/mode/solver.py

```diff
@@ -268,41 +286,32 @@ def conductivity_model_for_pec(eps, threshold=0.9 * pec_val):
             mat_dtype = np.complex128 if mat_precision == "double" else np.complex64
         else:
             # 1) check if complex or not
-            complex_solver = False
-            if (
-                cls.isinstance_complex(eps_tensor)
-                or cls.isinstance_complex(mu_tensor)
+            complex_solver = (
+                cls.isinstance_complex(eps)
+                or cls.isinstance_complex(mu)
                 or np.any([cls.isinstance_complex(f) for f in der_mats])
-            ):
-                complex_solver = True
+            )
             # 2) determine precision
             if complex_solver:
                 mat_dtype = np.complex128 if mat_precision == "double" else np.complex64
             else:
                 if mat_precision == "double":
                     mat_dtype = np.float64
 
-        # data type conversion
-        eps_tensor = cls.type_conversion(eps_tensor, mat_dtype)
-        mu_tensor = cls.type_conversion(mu_tensor, mat_dtype)
-        der_mats = [cls.type_conversion(f, mat_dtype) for f in der_mats]
-        neff_guess = cls.type_conversion(np.array([neff_guess]), mat_dtype)[0]
+        return mat_dtype
 
-        # initial vector for eigensolver
-        vec_init = cls.set_initial_vec(Nx, Ny, mat_dtype=mat_dtype, is_tensorial=is_tensorial)
-
-        # call solver
-        if is_tensorial:
-            return cls.solver_tensorial(
-                eps_tensor, mu_tensor, der_mats, num_modes, vec_init, neff_guess
-            )
-
-        return cls.solver_diagonal(eps_tensor, mu_tensor, der_mats, num_modes, vec_init, neff_guess)
+    @classmethod
+    def trim_small_values(cls, mat: sp.csr_matrix, tol: float) -> sp.csr_matrix:
+        """Eliminate elements of matrix ``mat`` for which ``abs(element) / abs(max_element) < tol``.
+        This operates in-place on mat so there is no return.
+        """
+        max_element = np.amax(np.abs(mat))
+        mat.data *= np.abs(mat.data) / max_element > tol
```

Comment:
Will it be safer with `mat.data *= np.logical_or(np.abs(mat.data) / max_element > tol, np.abs(mat.data) > tol)`. This doesn't change the number of nonzeros in your example.

Reply:
Good thinking, I updated this and added a line to changelog, looks good now?

Reply:
All look good to me